MovingBlocks/Terasology
# Add Server Blacklist and Whitelist

### Contains

fix issue #1608.

#### Summary

Implementation for server whitelisting/blacklisting. Netty doesn’t allow refusing connections, but it is possible to immediately close a connection after opening one. I do this by making the network system force disconnect the client if the client ID is found on the blacklist. The client is notified when they are blocked by a new message in the scheduleTimeout function. To make it easier to communicate if the client is on a blacklist or not, protobuf ServerInfoMessage has the blacklist/whitelist added.

#### Details

NetworkSystemImpl is the location where the server rejects or accepts the client. The blacklist file is found and if the client’s ID is found on the list, the client is immediately force disconnected.

For the protobuf changes, the only new thing is that an error message is sent across. Changes in NetworkSystemImpl getServerInfoMessage method provide the data for the new protobuf change.

ClientConnectionHandler is changed so that the user can receive feedback about whether or not they have been blacklisted. The scheduleTimeout function was moved to after the NetMessage was declared. It looks to me that everything still works correctly, and we need it there to access the blacklist. The NetMessage needs to be initialized so we can tell the client that they have been blacklisted, otherwise it just tells them that the server timed out, which is a confusing error message.

The new class doesn’t do much at the moment, but it could be accessed easily from some other source, such as the server facade.

### How to test

At the moment, the blacklist appears to work.

1. Run the game once to generate the blacklist/whitelist files.

2. Add your client ID to the blacklist in JSON format. For example, my blacklist.json contains the line:

<code> ["6a5f11f7-4038-4ef0-91ac-86cb957588b1","01264d12-27cf-4699-b8e2-bdc92ac8ef73"] </code>

These two IDs are for my first and second clients. To find your own client ID, you can run the code with a normal client and headless client, and look at the testing print statement that says “ID: \<client ID\>.”

3. Try to connect to a server and observe how it says that the client is on the blacklist. Also try to connect with another client and notice how nothing is changed. A similar process can be done for testing the whitelist

### Outstanding before merging

- [x] Implementation for the whitelist needs to be completed.
- [ ] Allow users to be blacklisted/whitelisted by IP address.
- [ ] Need to add some unit tests.
- [ ] Add javadocs
- [ ] Will need translation work
